src/reflect/scala/reflect/internal/Symbols.scala

```diff
@@ -2386,11 +2386,23 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
       } else Nil
     }
 
+    private[this] var isOverridingSymbolCache = 0
+
     /** Equivalent to allOverriddenSymbols.nonEmpty, but more efficient. */
-    lazy val isOverridingSymbol = (
+    private def computeIsOverridingSymbol: Boolean = (
          canMatchInheritedSymbols
       && owner.ancestors.exists(base => overriddenSymbol(base) != NoSymbol)
     )
+    final def isOverridingSymbol: Boolean = {
+      val curRunId = currentRunId
+      if (isOverridingSymbolCache == curRunId) true
+      else if (isOverridingSymbolCache == -curRunId) false
+      else {
+        val result = computeIsOverridingSymbol
+        isOverridingSymbolCache = (if (result) 1 else -1) * curRunId
+        result
+      }
```

Comment:
While this fixes a bug in the cache under multi-run compilation, we might still have a bug here if the result of this method is phase dependent. I'm thinking of a case where something doesn't override at typer, but does override at erasure.

Surveying the the call heirarchy:

`allOverridingSymbols` uses this, and it has hot callsites in `typedBlock => isPossibleInRefinement`; and in `SpecializeTypes.needsSpecialOverride`.

`nextOverridingSymbol` also uses this, and it has hot callsites in `Namer.methodSig` (for return type and default method inheritance).

So we wouldn't want to drop the caching altogether, but we could bypass the cache after we get to a certain phase.

To motivate this more precisely, we could try to craft an example that is vulnerable to the bug.

Reply:
Concrete examples.

## Incorrect caching across runs.

Prior to this PR:
```
⚡ scala-pr 6197
scala> :power

scala> :paste -raw

package p1; class A; class B extends A { def foo = 42 }

scala> typeOf[p1.B].decl(TermName("foo")).isOverridingSymbol
res0: Boolean = false

scala> :paste -raw

package p1; class A { def foo = 42 }

scala> typeOf[p1.B].decl(TermName("foo")).isOverridingSymbol
res1: Boolean = false
```

After, we correctly recompute the override status of `B.a` after the parent class was separately redefined.

```
⚡ scala-pr 6197
scala> :power

scala> :paste -raw

package p1; class A; class B extends A { def foo = 42 }

scala> typeOf[p1.B].decl(TermName("foo")).isOverridingSymbol
res0: Boolean = false

scala> :paste -raw

package p1; class A { def foo = 42 }

scala> typeOf[p1.B].decl(TermName("foo")).isOverridingSymbol
res1: Boolean = true
```

## Incorrect caching across phases

This problem has _not_ been addressed by this PR.

```scala
trait A[X] { def a(x: X) }; trait B extends A[String] { def a(s: String) }

scala> def B_a = typeOf[B].decl(TermName("a"))
B_a: $r.intp.global.Symbol

```

`exitingErasure(B_a.isOverridingSymbol)` should give `false`, but will still yield `true` if `isOverridingSymbol` was called pre-erasure:


```
scala> { println(exitingErasure(B_a.isOverridingSymbol)) }
false
```

```
scala> { println(B_a.isOverridingSymbol); println(exitingErasure(B_a.isOverridingSymbol)) }
true
true
```

I propose to leave this problem as TODO.

Reply:
Ok, but do add the `// TODO`